### PR TITLE
Add customizable accessoryView besides messageContainer

### PR DESCRIPTION
**Pull request by hyouuu**
What does this implement/fix? Explain your changes.
---------------------------------------------------
Add an accessoryView and give client a chance to config it to whatever they like (e.g. a button, image etc)

Does this close any currently open issues?
------------------------------------------
And I found one ;) https://github.com/MessageKit/MessageKit/issues/190

Any relevant logs, error output, etc?
-------------------------------------
NA

Any other comments?
-------------------
Let me know how you want to approach it!

Where has this been tested?
---------------------------
**Devices/Simulators:** iPhone 8+ simulator

**iOS Version:** 11

**Swift Version:** 4

**MessageKit Version:** 0.12

Example screenshot where in configCell added a UIView with green background:

screen shot 2017-12-31 at 1 01 22 pm

### CODE REVIEW

**Comment by SD10** [comment]
@hyouuu Thanks for submitting this! This is definitely on the todo list but it could take me some time to review. This could be more difficult than it seems because it changes so much in `MessagesCollectionViewFlowLayout`. An accessory view limits the available space for things like `messageContainerView`, `cellTopLabel`, cellBottomLabel`, depending on its location.

**Comment by hyouuu** [comment]
Definitely! I'll play around long message & labels etc and see where it can be improved; I'm in need of this so just made the basic version - will try to cover corner cases as I get to it and add to this PR ;)

**Comment by hyouuu** [comment]
Long piece seems to be working fine as is o_O
screen shot 2017-12-31 at 12 14 20 pm

**Comment by SD10** [comment]
@hyouuu It will definitely require changes to the methods:
`messageContainerMaxWidth`, `cellBottomLabelMaxWidth`, `cellTopLabelMaxWidth`, etc.

**Comment by hyouuu** [comment]
Yep I'm working on them as we speak - commit coming soon

File: Sources/Protocols/MessagesDataSource.swift [hunk 1]
```diff
@@ -81,6 +81,14 @@ public protocol MessagesDataSource: AnyObject {
...
     func cellBottomLabelAttributedText(for message: MessageType, at indexPath: IndexPath) -> NSAttributedString?
 
+    /// Additional config for a cell, called in cellForItem method. e.g. customize accessoryView.   
+    ///
+    /// - Parameters:
+    ///   - cell: The `MessageCollectionViewCell` that can be configured.
+    ///   - indexPath: The `IndexPath` of the cell.
+    ///
+    /// The default value returned by this method is `nil`.
+    func additionalConfig(for cell: MessageCollectionViewCell, at indexPath: IndexPath)
```
Comment (SD10) [inline Sources/Protocols/MessagesDataSource.swift]: I'm looking into this, but we may not want to add it in this PR. Let's keep the scope small

**Comment by SD10** [comment]
I think this resembles #418. Figuring out how to allow users to customize the `AvatarView` and what kind of `AccessoryView` is a similar problem.

File: Sources/Protocols/MessagesDataSource.swift (hunk 1 above)
Comment (hyouuu) [reply Sources/Protocols/MessagesDataSource.swift]: What's your suggestion on letting clients config the accessoryView? A specific setupAccessoryView delegate?

**Comment by hyouuu** [comment]
Adjusted messageContainerMaxWidth and did some refactor e.g. made config more generic with messageType.

Seems to work fine with padding and long messages:

Message without accessoryView (size & padding both .zero):
screen shot 2017-12-31 at 12 58 13 pm

With accessoryView (size 36 & padding 14):
screen shot 2017-12-31 at 1 01 22 pm

**Comment by hyouuu** [comment]
For #418, I think we could supply a switch on whether to auto config avatarView, and if it's off, people can use `messagesDataSource.configCell(cell, for: message, at: indexPath)
` to add subviews to the plain avatarView to customize it

**Comment by mmuharem** [comment]
I like this! How would you add a delegate for it being tapped just as the other parts of the cell (toplabel, bottom label, avatar, etc..)?

**Comment by hyouuu** [comment]
@mmuharem how this would work will be like a UITableViewCell's contentView, which serves as a view holder and you could add anything on it (e.g. a UIButton with a selector upon tapping ;)

**Comment by hyouuu** [comment]
I could update the example code as soon as @SD10 thinks this is the right approach, but basically in the configCell delegate method you could do something like:
```
        let button = UIButton()
        button.frame = cell.accessoryView.bounds
        cell.accessoryView.addSubview(button)

        // This is my customized tapBlock, but you could use addTarget
        button.tapBlock = {
            // Do something
        }
```

**Comment by mmuharem** [comment]
@hyouuu okay, so the part that I have a question particularly is the 

```
// This is my customized tapBlock, but you could use addTarget
        button.tapBlock = {
            // Do something
        }
```

as this would mean that you have a selector in an extension. I am using the addTarget but this is the part I was not able to figure out. 

So, where and how did you put the .tapBlock function?

**Comment by hyouuu** [comment]
What's the issue with a function in an extension? You could put the method in the main class as well. Remember to add @objc if it's a selector func

**Comment by mmuharem** [comment]
Well, that's what I am trying to figure out. 

Here is what I've tried in the MessagesDataSource.swift file. Is this what you mean by the 'main class'?:

```
public protocol MessagesDataSource: AnyObject {
      ....
      @objc func toggleReact(_ button: UIButton)
}

public extension MessagesDataSource {
     ....
    func configCell...{
             button.addTarget(self, action:  #selector(self.toggleReact) , for: .touchUpInside)
      }

   func toggleReact(_ button: UIButton) {
        print("button Tapped")
    }

}

```

The errors I receive:
- button.addTarget (Argument of selector refers to instance method 'toggleReact'that is not exposed to objective c. Add objc to expose this instance
- @objc func toggleReact (@objc can only be used with members of classes , @objc protocols, and concrete extensions of classes. Replace @objc with "". 

Instead, if I use MessagesCollectionView.swift as the main class, then it compiles but crashes upon click with unrecognized selector

**Comment by hyouuu** [comment]
Well @mmuharem what I do is that I have a subclass of MessagesViewController called ChatVC that implements configCell as well as toggleReact:
```
class ChatVC: MessagesViewController,
    MessagesDataSource,
    MessagesDisplayDelegate,
    MessagesLayoutDelegate,
    MessageCellDelegate
{
    func configCell...{
        button.addTarget(self, action:  #selector(self.toggleReact) , for: .touchUpInside)
    }

    func toggleReact(_ button: UIButton) {
        print("button Tapped")
    }
}
```

**Comment by mmuharem** [comment]
omg so simple that I overlooked that. Thank you so much for the help!

**Comment by SD10** [comment]
@hyouuu To update you on this it's being blocked by some other things 😐 I need to do some stuff inside of `MessagesCollectionViewFlowLayout`

**Comment by hyouuu** [comment]
Thanks for the update - hope everything goes well!

**Comment by SD10** [comment]
@hyouuu Tonight I'm going to break most of the internals for `MessagesCollectionViewFlowLayout`. We're going to have to re-open this another time. It may not be possible to resolve the merge conflicts, and this PR is great, but the logic is not correct (yet). I'll use this as a basis for when we make the addition and credit you for any work I use or migrate it for you and ping you for assistance 😃 Thanks for taking the time to put this together

**Comment by hyouuu** [comment]
@SD10 thanks for the update and the changes sound exciting! Hopefully we can get to this soon as otherwise I wouldn't be able to update MessageKit for my current in-dev app T_T Let me know of any assist I could help provide!

**Comment by SD10** [comment]
We can move forward on this again but it really does take a lot of work to get in. 
It would affect the following methods:
```
cellContentHeight(for message: MessageType, at indexPath: IndexPath) -> CGFloat
cellTopLabelMaxWidth(for message: MessageType, at indexPath: IndexPath) -> CGFloat
cellBottomLabelMaxWidth(for message: MessageType, at indexPath: IndexPath) -> CGFloat
messageContainerMaxWidth(for message: MessageType, at indexPath: IndexPath) -> CGFloat
```

Things to consider: 
- What if `AvatarView` width + `AccessoryView` width > `itemWidth`?

Pretty extreme edge case but some stuff like this is on my mind.

**Comment by hyouuu** [comment]
Cool will give a try soon. For the edge case, my first thought is `AccessoryView` is more important than AvatarView because it usually takes some kind of user action rather than pure displaying (although you could argue `AvatarView` could be customized to do actions too).
What if `AvatarView` width > `itemWidth` or `AccessoryView` width > `itemWidth`? Should we display part of them, shrink them or throw an error?

**Pull request closed without merging**